src/frequenz/client/dispatch/test/_service.py

```diff
@@ -49,18 +55,59 @@ class FakeService:
     _shuffle_after_create: bool = False
     """Whether to shuffle the dispatches after creating them."""
 
+    def _check_access(self, metadata: grpc.aio.Metadata) -> None:
+        """Check if the access key is valid.
+
+        Args:
+            metadata: The metadata.
+
+        Raises:
+            grpc.RpcError: If the access key is invalid.
+        """
+        # metadata is a weird tuple of tuples, we don't like it
```

Comment:
> [!WARNING]
> A weird tuple of tuples is not the same as a dict.

A tuple of tuples is more like a multi-dict, where a key can have multiple values (the same happens with Mail/HTTP headers or URL query parameters). By just converting it to a dict blindly, you are getting a random value if a key has more than one value.

If you don't want to support multiple values for a key, I would get the last value for each key instead, I think it is more deterministic and probably more correct. Something like [`cgi.FieldStorage.getfirst`](https://docs.python.org/3/library/cgi.html#cgi.FieldStorage.getfirst) and [`cgi.FieldStorage.getlist`](https://docs.python.org/3/library/cgi.html#cgi.FieldStorage.getlist).

This will be probably needed by all APIs, so it might be worth putting into `client-base`.

Reply:
I doubt any client has to do this, given that this is the fake service...

Reply:
> By just converting it to a dict blindly, you are getting a random value if a key has more than one value.

I am not worried about that, I don't think we need or want to test that case in our tests.. (which is what this is for)

Reply:
Also missed it was only the test class. :innocent: